feat: use count api for documents count when fetching indices

damian-polewski opened the pull request:
## Summary

Closes #239103

This PR updates the `fetchIndices` function to use `_count` API instead of `_stats` API for documents count in Index Management plugin.

### Checklist

Check the PR satisfies following conditions. 

Reviewers should verify this PR satisfies this list as well.

- [ ] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/src/platform/packages/shared/kbn-i18n/README.md)
- [ ] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [ ] If a plugin configuration key changed, check if it needs to be allowlisted in the cloud and added to the [docker list](https://github.com/elastic/kibana/blob/main/src/dev/build/tasks/os_packages/docker_generator/resources/base/bin/kibana-docker)
- [ ] This was checked for breaking HTTP API changes, and any breaking changes have been approved by the breaking-change committee. The `release_note:breaking` label should be applied in these situations.
- [ ] [Flaky Test Runner](https://ci-stats.kibana.dev/trigger_flaky_test_runner/1) was used on any tests changed
- [x] The PR  description includes the appropriate Release Notes section, and the correct `release_note:*` label is applied per the [guidelines](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)
- [x] Review the [backport guidelines](https://docs.google.com/document/d/1VyN5k91e5OVumlc0Gb9RPa3h1ewuPE705nRtioPiTvY/edit?usp=sharing) and apply applicable `backport:*` labels.

damian-polewski committed:
feat: use count api for documents count when fetching indices

SoniaSanzV reviewed (commented):
Code changes LGTM, can we confirm that don't need to also replace `_metering/stats` with that?

damian-polewski committed:
Merge branch 'main' into index_management/use_count_api_for_docs_count

damian-polewski committed:
Merge branch 'main' into index_management/use_count_api_for_docs_count

damian-polewski committed:
Merge branch 'main' into index_management/use_count_api_for_docs_count

damian-polewski committed:
Merge branch 'main' into index_management/use_count_api_for_docs_count

damian-polewski committed:
fix: _metering_stats documents count

damian-polewski commented:
Hey @SoniaSanzV, good catch! I confirmed that we also should changed it so I replaced the documents count using `_metering_stats` with `_count`. Updated in https://github.com/elastic/kibana/pull/241357/commits/b026e55d834af1ccb827f59875e1395b056bf2a0

damian-polewski committed:
Merge branch 'main' into index_management/use_count_api_for_docs_count

SoniaSanzV commented on x-pack/platform/plugins/shared/index_management/server/lib/fetch_indices.ts:
```diff
@@ -56,42 +56,52 @@ async function fetchIndicesCall(
+          .count({
+            index: indexName,
+          })
+          .catch(() => ({ count: 0 }));
```
why return `count:0` instead of an error? I'm worry that this would hide real issues

SoniaSanzV replied:
In that case, I think your approach of returning `count:0` is better than fallback to `_stats `

SoniaSanzV commented on x-pack/platform/plugins/shared/index_management/server/lib/fetch_indices.ts:
```diff
@@ -56,42 +56,52 @@ async function fetchIndicesCall(
-          primary_size: new ByteSizeValue(
-            indexStats?.primaries?.store?.size_in_bytes ?? 0
-          ).toString(),
+    return await Promise.all(
```
I don't love that we need to fetch the count for every single index, there are users with many many indices. But from the api docs I see that this endpoint returns a summary of the indices docs if not fetched individually. If this become a problem, maybe we'll need to implement pagination.

damian-polewski replied:
Exactly, unfortunately the [count API](https://www.elastic.co/docs/api/doc/elasticsearch/operation/operation-count) doesn't support fetching document count for multiple indices because it returns a sum of documents so that's why I had to fetch the document count individually (I used `Promise.all()` to resolve promises concurrently). The another issue is that the [GET indices operation](https://www.elastic.co/docs/api/doc/elasticsearch/operation/operation-indices-get) doesn't support pagination so we are not able to apply it here. 
I'm not sure what could be the alternative approach considering cases of users with a LOT of indices. Maybe  @mattkime could take a look and share his perspective?

Why do we want to replace the stats API with count API for document count:
> The `_stats` API uses Lucene document counts, which includes nested documents. Vectors are stored as nested documents, which means that the docs count from that API is at least double what a vector search user would expect to see. The `_count` API only includes root documents, which is more natural for users to understand.

For example if I ingest 5 documents with semantic text fields...
```
PUT semantic_test_data
{
  "mappings": {
    "properties": {
      "content": {
        "type": "semantic_text"
      }
    }
  }
}

POST semantic_test_data/_bulk
{ "index": { "_id": "1" } }
{ "content": "Elasticsearch is a distributed, RESTful search and analytics engine capable of solving a growing number of use cases." }
{ "index": { "_id": "2" } }
{ "content": "Kibana lets you visualize your Elasticsearch data and navigate the Elastic Stack." }
{ "index": { "_id": "3" } }
{ "content": "The semantic_text field type enables semantic search by automatically generating embeddings for your text." }
{ "index": { "_id": "4" } }
{ "content": "You can use Elastic's ELSER model for semantic search without managing inference pipelines." }
{ "index": { "_id": "5" } }
{ "content": "Semantic search finds results based on meaning, not just keywords." }
```
...when using `_stats API` I can see 10 documents...
 
<img width="1642" height="405" alt="Screenshot 2025-12-01 at 14 24 52" src="https://github.com/user-attachments/assets/6dae0086-6ebc-4160-87b7-59f7a95fd59e" />

...but, when using `_count API` it returns 5 documents properly.

<img width="1644" height="398" alt="Screenshot 2025-12-01 at 14 30 33" src="https://github.com/user-attachments/assets/3f80b0f7-bf0f-41bc-9d49-0cfcd01c5c73" />

mattkime replied:
@damian-polewski I think we should talk to the ES engineers to see what they might have to offer. Its pretty clear to me that the count api isn't the swap in replacement that both teams were hoping it would be.

SoniaSanzV commented on x-pack/platform/plugins/shared/index_management/server/lib/fetch_indices.test.ts:
```diff
@@ -125,6 +130,7 @@ describe('[Index management API Routes] fetch indices lib function', () => {
         }),
       });
+      getCount.mockResolvedValue({ count: 1 });
       getIndicesStats.mockResolvedValue({
```
Can you add some tests cases where the count api and the _stats and _metering/stats return different count values? To cover the case that this pr is implementing

damian-polewski committed:
Merge branch 'main' into index_management/use_count_api_for_docs_count

damian-polewski committed:
Merge branch 'main' into index_management/use_count_api_for_docs_count

damian-polewski committed:
Merge branch 'main' into index_management/use_count_api_for_docs_count

damian-polewski committed:
fix: cr changes

elasticmachine commented:
## :broken_heart: Build Failed
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/369887)
* Commit: 57562d1c61e51bc54ad94a4ce32f2b38d926ae8b

### Failed CI Steps
* [Check Types](https://buildkite.com/elastic/kibana-pull-request/builds/369887#019afae6-c844-4ab5-8afb-d232fdad15f9)

### History
* :green_heart: [Build #368525](https://buildkite.com/elastic/kibana-pull-request/builds/368525) succeeded 48bef750a4e6c4139c818054ee942f8d1203ae39
* :green_heart: [Build #367133](https://buildkite.com/elastic/kibana-pull-request/builds/367133) succeeded 9b6ede518f97cb3413c24fe2e9d966d263270cf6
* :yellow_heart: [Build #358916](https://buildkite.com/elastic/kibana-pull-request/builds/358916) was flaky b026e55d834af1ccb827f59875e1395b056bf2a0
* :green_heart: [Build #358034](https://buildkite.com/elastic/kibana-pull-request/builds/358034) succeeded bd550bca7bc734b38c7e99be8e948ef9420db389

cc @damian-polewski

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/369887","number":369887,"commit":"57562d1c61e51bc54ad94a4ce32f2b38d926ae8b"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/368525","number":368525,"commit":"48bef750a4e6c4139c818054ee942f8d1203ae39"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/367133","number":367133,"commit":"9b6ede518f97cb3413c24fe2e9d966d263270cf6"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/358916","number":358916,"commit":"b026e55d834af1ccb827f59875e1395b056bf2a0"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/358034","number":358034,"commit":"bd550bca7bc734b38c7e99be8e948ef9420db389"}],"number":369887}
buildkite-pr-comment-->

damian-polewski committed:
fix: add missing logger mock
